src/cmap/auth/mongodb_oidc.ts

```diff
@@ -1,39 +1,136 @@
 }
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export interface OIDCRequestTokenResult {
   accessToken: string;
   expiresInSeconds?: number;
   refreshToken?: string;
 }
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export type OIDCRequestFunction = (
-  idl: OIDCMechanismServerStep1
+  principalName: string,
+  serverResult: OIDCMechanismServerStep1,
+  timeout: AbortSignal | number
 ) => Promise<OIDCRequestTokenResult>;
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export type OIDCRefreshFunction = (
-  idl: OIDCMechanismServerStep1,
-  result: OIDCRequestTokenResult
+  principalName: string,
+  serverResult: OIDCMechanismServerStep1,
+  result: OIDCRequestTokenResult,
+  timeout: AbortSignal | number
 ) => Promise<OIDCRequestTokenResult>;
+
+/** @internal */
+export const OIDC_WORKFLOWS = {
+  callback: new CallbackWorkflow(),
+  aws: new AwsDeviceWorkflow(),
+  azure: undefined,
+  gcp: undefined,
+  // Set the protype to null to avoid false returns and keep types.
+  __proto__: null
+};
+
+/**
+ * OIDC auth provider.
+ */
+export class MongoDBOIDC extends AuthProvider {
+  /**
+   * Instantiate the auth provider.
+   */
+  constructor() {
+    super();
+  }
+
+  /**
+   * Authenticate using OIDC
+   */
+  override auth(authContext: AuthContext, callback: Callback): void {
+    const { connection, credentials, response } = authContext;
+
+    if (response && response.speculativeAuthenticate) {
+      return callback();
+    }
+
+    if (!credentials) {
+      return callback(new MongoMissingCredentialsError('AuthContext must provide credentials.'));
+    }
+
+    getWorkflow(credentials, (error, workflow) => {
+      if (error || !workflow) {
+        return callback(error);
+      }
+      workflow
+        .execute(connection, credentials)
+        .then(result => {
+          return callback(undefined, result);
+        })
+        .catch(error => {
+          callback(error);
+        });
+    });
+  }
+
+  /**
+   * Add the speculative auth for the initial handshake.
+   */
+  override prepare(
+    handshakeDoc: HandshakeDocument,
+    authContext: AuthContext,
+    callback: Callback<HandshakeDocument>
+  ): void {
+    const { credentials } = authContext;
+
+    if (!credentials) {
+      return callback(new MongoMissingCredentialsError('AuthContext must provide credentials.'));
+    }
+
+    getWorkflow(credentials, (error, workflow) => {
+      if (error || !workflow) {
+        return callback(error);
+      }
+      workflow
+        .speculativeAuth()
+        .then(result => {
+          return callback(undefined, { ...handshakeDoc, ...result });
+        })
+        .catch(error => {
+          callback(error);
+        });
+    });
+  }
+}
+
+/**
+ * Gets either a device workflow or callback workflow.
+ */
+function getWorkflow(credentials: MongoCredentials, callback: Callback<Workflow>): void {
```

Comment:
Is there any reason for a callback in this method?  There's no async work, so we could directly throw / return

Reply:
The code got really odd when combining the callback from the caller and potentially throwing from this method. So the caller had both a try/catch and a callback call with the error. The combination of the two just seemed weird so I stuck with one - but happy to change if you feel strongly about it.

Reply:
That's a good point, I don't feel strongly.

A potential alternative is to make this async, and then just then-chain onto it where we call it.  Then, the error this function rejects with will be caught by the `catch` handler for the promise chain.

Reply:
I think I'd just prefer to keep it as-is, and then once we come back when async/awaiting all the internals just make it async then.